include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,transmitted,V,L,N,NdotL,VdotL,orientedEta,rcpOrientedEta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(T,H);
-      retval.BdotH = dot(B,H);
-    }
-    return valid;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static bool compute(
-    out AnisotropicMicrofacetCache retval,
-    const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,interaction,_sample,eta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(interaction.T,H);
-      retval.BdotH = dot(interaction.B,H);
-    }
-    return valid;
-  }
-
-  float TdotH;
-  float BdotH;
+    using this_t = SIsotropicMicrofacetCache<T>;
+    using scalar_type = T;
+    using vector3_type = vector<scalar_type, 3>;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+
+    using ray_dir_info_type = ray_dir_info::SBasic<scalar_type>;
+    using isotropic_type = surface_interactions::SIsotropic<ray_dir_info_type>;
+    using sample_type = SLightSample<ray_dir_info_type>;
+
+    // always valid because its specialized for the reflective case
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL, NBL_REF_ARG(scalar_type) LplusV_rcpLen)
+    {
+        LplusV_rcpLen = rsqrt<scalar_type>(2.0 + 2.0 * VdotL);
+
+        this_t retval;
+
+        retval.VdotH = LplusV_rcpLen * VdotL + LplusV_rcpLen;
+        retval.LdotH = retval.VdotH;
+        retval.NdotH = (NdotL + NdotV) * LplusV_rcpLen;
+        retval.NdotH2 = retval.NdotH * retval.NdotH;
+
+        return retval;
+    }
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL)
+    {
+        float dummy;
+        return createForReflection(NdotV, NdotL, VdotL, dummy);
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static this_t createForReflection(
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample)
+    {
+        return createForReflection(interaction.getNdotV(), _sample.getNdotL(), _sample.getVdotL());
+    }
+    // transmissive cases need to be checked if the path is valid before usage
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        const bool transmitted, NBL_CONST_REF_ARG(vector3_type) V, NBL_CONST_REF_ARG(vector3_type) L,
+        NBL_CONST_REF_ARG(vector3_type) N, const scalar_type NdotL, const scalar_type VdotL,
+        const scalar_type orientedEta, const scalar_type rcpOrientedEta, NBL_REF_ARG(vector3_type) H
```

Comment:
I have a slightly better idea, instead of having a pseudo-hidden constructor/initializer we make a `create` and add an `operator bool() CONST_METHOD {return isnan(NdotH);}`

Reply:
also take the `fresnel::OrientedEtas<scalar_type>` struct

Reply:
you can however make a `create(const bool transmitted, const vector_type V, const vector_type L, REF_ARG(vector_type) H)`

Reply:
continued in

https://github.com/Devsh-Graphics-Programming/Nabla/pull/899/files?diff=unified&w=1#r2191004375 

https://github.com/Devsh-Graphics-Programming/Nabla/pull/899/files?diff=unified&w=1#r2191009537

https://github.com/Devsh-Graphics-Programming/Nabla/pull/899/files?diff=unified&w=1#r2191017316